src/main/scala/epic/dense/CachingLookupAffineWithSparse.scala

```diff
@@ -0,0 +1,131 @@
+
+import scala.runtime.ScalaRunTime
+import breeze.linalg._
+import epic.features.SegmentedIndex
+import epic.framework.Feature
+import scala.collection.mutable.HashMap
+import scala.util.Random
+
+/**
+ * Used at the input layer to cache lookups and 
+ */
+case class CachingLookupAffineWithSparse[FV](numOutputs: Int,
+                                             numWordInputs: Int,
+                                             word2vecIndexed: Word2VecIndexed[String],
+                                             auxInputDomainSizes: Array[Int],
+                                             includeBias: Boolean = true) extends Transform[Array[Int], DenseVector[Double]] {
+ 
+  val numAuxInputs = auxInputDomainSizes.size
+  val totalAuxInputEntries = auxInputDomainSizes.foldLeft(0)(_ + _)
+  val numInputs = numWordInputs * word2vecIndexed.wordRepSize + totalAuxInputEntries
+  
+  val index = new AffineTransform.Index(numOutputs, numInputs, includeBias)
+  
+  def extractLayer(weights: DenseVector[Double], forTrain: Boolean) = {
+    val mat = weights(0 until (numOutputs * numInputs)).asDenseMatrix.reshape(numOutputs, numInputs, view = View.Require)
+    val bias = if(includeBias) {
+      weights(numOutputs * numInputs until index.size)
+    } else {
+      DenseVector.zeros[Double](numOutputs)
+    }
+    new Layer(mat, bias)
+  }
+  
+  def initialWeightVector(initWeightsScale: Double, rng: Random, outputLayer: Boolean, spec: String) = {
+    val myWeights = if (outputLayer) {
+      DenseVector(Array.tabulate(index.size)(i => 0.0))
+    } else if (spec == "magic") {
+      AffineTransform.getMagicAffineWeights(index.size, numInputs, numOutputs, initWeightsScale, rng)
+    } else {
+      AffineTransform.getGaussianAffineWeights(index.size, initWeightsScale, rng)
+    }
+    myWeights
+//    DenseVector(Array.tabulate(index.size)(i => if (!outputLayer) rng.nextGaussian * initWeightsScale else 0.0))
+  }
+  
+  def clipHiddenWeightVectors(weights: DenseVector[Double], norm: Double, outputLayer: Boolean) {
+    if (!outputLayer) {
+      AffineTransform.clipHiddenWeightVectors(numOutputs, numInputs, weights, norm)
+    }
+  }
+  
+  def getInterestingWeightIndicesForGradientCheck(offset: Int): Seq[Int] = {
+    (offset until offset + Math.min(10, index.size))
+  }
+
+  case class Layer(weights: DenseMatrix[Double], bias: DenseVector[Double]) extends Transform.Layer[Array[Int],DenseVector[Double]] {
+    
+    override val index = CachingLookupAffineWithSparse.this.index
+    
+    val weightst = weights.t
+
+    // Cache stores pairs of (word identity, position) mapped to the final results of
+    // these being multiplied by the parameter vector. Note that although the same
+    // word vector is used for each word identity, the parameter vector depends
+    // on the position.
+//    val cache = new HashMap[(Int,Int),DenseVector[Double]]
+    val caches = Array.tabulate(numWordInputs)(i => new HashMap[Int,DenseVector[Double]])
```

Comment:
if you're still getting burned by HashMap overhead, you might consider OpenAddressHashArray which was designed for more or less this purpose